### PR TITLE
[llvm][arm] add T1 and T2 assembly options for vlldm and vlstm

### PR DESCRIPTION
**Pull request by sivan-shani**
T1 allow for an optional registers list,
the register list must be {d0-d15}.
T2 define a mandatory register list,
the register list must be {d0-d31}.

The requirements for T1/T2 are as follows:
                T1              T2
Require:        v8-M.Main,      v8.1-M.Main,
                secure state    secure state
16 D Regs       valid           valid
32 D Regs       UNDEFINED       valid
No D Regs       NOP             NOP

### CODE REVIEW

**Comment by rgwott** [comment]
LGTM, but I would prefer if you waited for someone else's review as well.

**Review by rgwott — APPROVED** [review]

File: llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp [hunk 1]
```diff
@@ -7030,3 +7033,26 @@ static DecodeStatus DecodeT2AddSubSPImm(MCInst &Inst, unsigned Insn,
...
+
+  const unsigned Rn = fieldFromInstruction(Insn, 16, 4);
+  // Adding Rn, holding memory location to save/load to/from, the only argument
+  // that being encoded.
+  // '$Rn' in the assembly.
+  if (!Check(S, DecodeGPRRegisterClass(Inst, Rn, Address, Decoder)))
+    return MCDisassembler::Fail;
+  // An optional predicate, '$p' in the assembly.
+  Inst.addOperand(MCOperand::createImm(0));
+  // A register that required for when the predicate exists (see function
```
Comment (jthackray) [inline llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp:7050]: Nit: s/that required/that is required/

File: llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp [hunk 2]
```diff
@@ -7030,3 +7033,26 @@ static DecodeStatus DecodeT2AddSubSPImm(MCInst &Inst, unsigned Insn,
...
+  // 'UpdateThumbVFPPredicate' and 'ARMInstrFormats.td::pred')
+  Inst.addOperand(MCOperand::createReg(ARM::NoRegister));
+  // Am immediate that represnts a floating point registers list. '$regs' in the
```
Comment (jthackray) [inline llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp:7053]: Nit: typo: s/represnts/represents/

**Comment by rgwott** [comment]
Looks like CI is breaking two test cases:
`Failed Tests (2):
  LLVM :: CodeGen/ARM/cmse-vlldm-no-reorder.mir
  LLVM :: CodeGen/ARM/vlldm-vlstm-uops.mir`

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp [hunk 3]
```diff
@@ -7596,6 +7601,38 @@ bool ARMAsmParser::validateInstruction(MCInst &Inst,
...
+  case ARM::VLLDM:
+    [[fallthrough]];
+  case ARM::VLLDM_T2:
+    [[fallthrough]];
+  case ARM::VLSTM:
+    [[fallthrough]];
+  case ARM::VLSTM_T2: {
+    // Since in some cases both T1 and T2 are valid, tablegen can not always
+    // pick the correct instruction.
+    if (Operands.size() == 4) { // a register list has been provided
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:7613]: I think you will always have 4 operands, because you are doing the disassembly in `DecodeLazyLoadStoreMul`. This could be omitted, or made an assert.

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp [hunk 4]
```diff
@@ -7596,6 +7601,38 @@ bool ARMAsmParser::validateInstruction(MCInst &Inst,
...
+      ARMOperand &Op = static_cast<ARMOperand &>(
+          *Operands[3]); // the register list, a dpr_reglist
+      if (Op.isDPRRegList()) {
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:7616]: This seems like it should be an assert rather than an `if`. Can this ever not be the case?

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp [hunk 5]
```diff
@@ -7596,6 +7601,38 @@ bool ARMAsmParser::validateInstruction(MCInst &Inst,
...
+        auto &RegList = Op.getRegList();
+        // T2 requires v8.1-M.Main (cannot be handled by tablegen)
+        if (RegList.size() == 32 && !hasV8_1MMainline()) {
+          return Error(Op.getEndLoc(), "T2 version requires v8.1-M.Main");
+        }
+        // When target has 32 D registers, T1 is undefined.
+        if (hasD32() && RegList.size() != 32) {
+          return Error(Op.getEndLoc(), "operand must be exactly {d0-d31}");
+        }
+        // When target has 16 D registers, both T1 and T2 are valid.
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:7626]: Might be worth clarifying that T2 is a NOP in this case.

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp [hunk 6]
```diff
@@ -8731,6 +8768,34 @@ bool ARMAsmParser::processInstruction(MCInst &Inst,
   }
 
   switch (Inst.getOpcode()) {
+  case ARM::VLLDM:
+    [[fallthrough]];
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:8772]: Unnecessary fallthrough.

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp [hunk 7]
```diff
@@ -8731,6 +8768,34 @@ bool ARMAsmParser::processInstruction(MCInst &Inst,
...
+          TmpInst.addOperand(Inst.getOperand(1));
+          TmpInst.addOperand(Inst.getOperand(2));
+          TmpInst.addOperand(Inst.getOperand(3));
+          Inst = TmpInst;
+          return true;
+        }
+      }
+    }
+    return false;
+  }
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:8798]: Is the idea here that the instructions are always parsed as `VLLDM`/`VLSTM`, and never into the T2 versions, and therefore need to be fixed at this point?

File: llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp [hunk 8]
```diff
@@ -7030,3 +7033,26 @@ static DecodeStatus DecodeT2AddSubSPImm(MCInst &Inst, unsigned Insn,
...
+  DecodeStatus S = MCDisassembler::Success;
+
+  const unsigned Rn = fieldFromInstruction(Insn, 16, 4);
+  // Adding Rn, holding memory location to save/load to/from, the only argument
+  // that being encoded.
+  // '$Rn' in the assembly.
+  if (!Check(S, DecodeGPRRegisterClass(Inst, Rn, Address, Decoder)))
+    return MCDisassembler::Fail;
+  // An optional predicate, '$p' in the assembly.
+  Inst.addOperand(MCOperand::createImm(0));
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp:7049]: If these are predicated you'll need to parse the predicate properly, using `DecodePredicateOperand`. Please also add tests for some predicated versions.

File: llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp [hunk 9]
```diff
@@ -7030,3 +7033,26 @@ static DecodeStatus DecodeT2AddSubSPImm(MCInst &Inst, unsigned Insn,
...
+  // A register that required for when the predicate exists (see function
+  // 'UpdateThumbVFPPredicate' and 'ARMInstrFormats.td::pred')
+  Inst.addOperand(MCOperand::createReg(ARM::NoRegister));
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp:7052]: `DecodePredicateOperand` will add this register for you.

File: llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp [hunk 10]
```diff
@@ -7030,3 +7033,26 @@ static DecodeStatus DecodeT2AddSubSPImm(MCInst &Inst, unsigned Insn,
...
+  // Am immediate that represnts a floating point registers list. '$regs' in the
+  // assembly.
+  Inst.addOperand(MCOperand::createImm(0)); // Arbitrary value, has no effect.
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp:7055]: I know the value of the immediate is arbitrary because it is never used, but consider using `DecodeRegListOperand` anyway.

File: llvm/lib/Target/ARM/MCTargetDesc/ARMInstPrinter.cpp [hunk 11]
```diff
@@ -91,6 +91,38 @@ void ARMInstPrinter::printInst(const MCInst *MI, uint64_t Address,
...
+    return;
+  }
+  case ARM::VLSTM_T2: {
+    const MCOperand &Reg = MI->getOperand(0);
+    O << '\t' << "vlstm" << '\t';
+    printRegName(O, Reg.getReg());
+    O << ", "
+      << "{d0 - d31}";
+    return;
+  }
```
Comment (tmatheson-arm) [inline llvm/lib/Target/ARM/MCTargetDesc/ARMInstPrinter.cpp:125]: I think if you used a non-arbitrary value for the register list operand (which `DecodeRegListOperand` should give you) then you could avoid need for custom printing.

File: llvm/test/CodeGen/ARM/cmse-vlldm-no-reorder.mir [hunk 12]
```diff
@@ -77,10 +77,10 @@ body:             |
     frame-setup CFI_INSTRUCTION offset $lr, -4
     frame-setup CFI_INSTRUCTION offset $r7, -8
     renamable $r0 = t2MOVi32imm @g
-    renamable $r0 = t2LDRi12 killed renamable $r0, 0, 14 /* CC::al */, $noreg :: (dereferenceable load (s32) from `float ()** bitcast (float (...)** @g to float ()**)`)
+    renamable $r0 = t2LDRi12 killed renamable $r0, 0, 14 /* CC::al */, $noreg :: (dereferenceable load 4 from `float ()** bitcast (float (...)** @g to float ()**)`)
     tBLXNS_CALL killed renamable $r0, csr_aapcs, implicit-def dead $lr, implicit $sp, implicit-def $sp, implicit-def $s0
     renamable $r0 = t2MOVi32imm @a
-    VSTRS killed renamable $s0, killed renamable $r0, 0, 14 /* CC::al */, $noreg :: (store (s32) into @a)
+    VSTRS killed renamable $s0, killed renamable $r0, 0, 14 /* CC::al */, $noreg :: (store 4 into @a)
```
Comment (tmatheson-arm) [inline llvm/test/CodeGen/ARM/cmse-vlldm-no-reorder.mir:83]: Are these changes related to the patch?

File: llvm/test/CodeGen/ARM/vlldm-vlstm-uops.mir [hunk 13]
```diff
@@ -1,8 +1,7 @@
 # RUN: llc -run-pass=if-converter %s -o - | FileCheck %s
 --- |
   target triple = "thumbv8m.main-arm-none-eabi"
-
-  define hidden void @foo(ptr nocapture %baz) local_unnamed_addr #0 {
+define hidden void @foo(void ()* nocapture %baz) local_unnamed_addr #0 {
```
Comment (tmatheson-arm) [inline llvm/test/CodeGen/ARM/vlldm-vlstm-uops.mir:4]: Please undo the whitespace change.

File: llvm/lib/Target/ARM/ARMPredicates.td [hunk 14]
```diff
@@ -183,6 +183,10 @@ def UseNegativeImmediates :
             AssemblerPredicate<(all_of (not FeatureNoNegativeImmediates)),
                                "NegativeImmediates">;
 
+def HasD32    : Predicate<"Subtarget->hasD32()">,
+                           AssemblerPredicate<(all_of FeatureD32),
+                           "32 D registers">;
+
```
[the remark this answers is not in the document]
Comment (sivan-shani) [reply llvm/lib/Target/ARM/ARMPredicates.td:189]: It is used (HasD16 was not used and was removed)

File: llvm/lib/Target/ARM/MCTargetDesc/ARMInstPrinter.cpp (hunk 11 above)
Comment (sivan-shani) [reply llvm/lib/Target/ARM/MCTargetDesc/ARMInstPrinter.cpp:125]: Seems as DecodeRegListOperand allow only for up to 16 registers.

File: llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp (hunk 10 above)
Comment (sivan-shani) [reply llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp:7055]: DecodeRegListOperand allow only for up to 16 registers.

Comment (tmatheson-arm) [reply llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp:7055]: The register list is not actually encoded, so `DecodeRegListOperand` can't be used.

File: llvm/lib/Target/ARM/ARMPredicates.td (hunk 14 above)
[the remark this answers is not in the document]
Comment (tmatheson-arm) [reply llvm/lib/Target/ARM/ARMPredicates.td:189]: Where?

[the remark this answers is not in the document]
Comment (sivan-shani) [reply llvm/lib/Target/ARM/ARMPredicates.td:189]: I thought that FeatureD32 using it, but I think it is actually the other way around.

File: llvm/test/CodeGen/ARM/cmse-vlldm-no-reorder.mir (hunk 12 above)
Comment (sivan-shani) [reply llvm/test/CodeGen/ARM/cmse-vlldm-no-reorder.mir:83]: Yes, both contain VLLDM and VLSTM and the changes have changed the output, hence the files had to be updated.

**Comment by sivan-shani** [comment]
Windows pipeline is failing on mysterious file that does not seems to be related to the change:
```
># error: command failed with exit status: 0xc0000005
# executed command: 'c:\ws\src\build\bin\filecheck.exe' 'C:\ws\src\flang\test\Fir\memory-allocation-opt.fir'
# .---command stderr------------
# | FileCheck error: '<stdin>' is empty.
# | FileCheck command line:  c:\ws\src\build\bin\filecheck.exe C:\ws\src\flang\test\Fir\memory-allocation-opt.fir
# `-----------------------------
# error: command failed with exit status: 2
```

File: llvm/lib/Target/ARM/ARMExpandPseudoInsts.cpp [hunk 15]
```diff
@@ -1468,15 +1468,24 @@ void ARMExpandPseudo::CMSESaveClearFPRegsV8(
   if (passesFPReg)
     assert(STI->hasFPRegs() && "Subtarget needs fpregs");
 
-  // Lazy store all fp registers to the stack.
+  if (passesFPReg)
+    assert(STI->hasFPRegs() && "Subtarget needs fpregs");
+
+  // Lazy store all fp registers to the stack
```
[the remark this answers is not in the document]
Comment (sivan-shani) [reply llvm/lib/Target/ARM/ARMExpandPseudoInsts.cpp:1471]: Done, in anew commit.
(Used '--amend', but github created new commit)

File: llvm/test/CodeGen/ARM/cmse-vlldm-no-reorder.mir (hunk 12 above)
Comment (sivan-shani) [reply llvm/test/CodeGen/ARM/cmse-vlldm-no-reorder.mir]: seems to gone after rebasing

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp (hunk 5 above)
Comment (sivan-shani) [reply llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:7623]: From the Arm ARM pseudo code it seems that T2 might work also when target has 16 D regs.
The only explicit NOP cases are when:
1. Secure Floating-point is not in use.
2. Floating-point Extension and MVE are not implemented.

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp (hunk 3 above)
Comment (sivan-shani) [reply llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:7610]: `DecodeLazyLoadStoreMul` is part of the Decoding process (Binary representation to MCInst), this code handle the Parsing (Assembly to MCInst).
When the assembly does not contain a registry list (e.g. legacy T1) there will be only 3 operands.

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp (hunk 7 above)
Comment (sivan-shani) [reply llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:8794]: Essentially yes, TableGen can not distinguish between `DPRRegList` of 16/32 registers.
As a default, it picks the 16 regs case (T1)

**Review by tommat01 — APPROVED** [review]
LGTM

**Comment by tmatheson-arm** [comment]
(Windows flang test failure was unrelated to this patch)

**Comment by RKSimon** [comment]
@tmatheson-arm @sivan-shani This is causing failures on EXPENSIVE_CHECKS builds - please can you take a look?
https://lab.llvm.org/buildbot/#/builders/16/builds/61858
